websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2022 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.websocket;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.api.WebSocketListener;
+import org.eclipse.jetty.websocket.api.WriteCallback;
+import org.eclipse.jetty.websocket.server.JettyServerUpgradeRequest;
+import org.eclipse.jetty.websocket.server.JettyServerUpgradeResponse;
+import org.eclipse.jetty.websocket.server.JettyWebSocketServerContainer;
+import org.kohsuke.MetaInfServices;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+@Restricted(NoExternalUse.class)
+@MetaInfServices(Provider.class)
+public class Jetty10Provider implements Provider {
+
+    private static final String ATTR_LISTENER = Jetty10Provider.class.getName() + ".listener";
+
+    // TODO does not seem possible to use HttpServletRequest.get/setAttribute for this
+    private static final Map<Listener, Session> sessions = new WeakHashMap<>();
+
+    private JettyWebSocketServerContainer container;
+
+    public Jetty10Provider() {
+        JettyWebSocketServerContainer.class.hashCode();
+    }
+
+    private synchronized void init(HttpServletRequest req) throws Exception {
+        if (container == null) {
+            // TODO JettyWebSocketServletContainerInitializer.configure(context, null) but where do we get the ServletContextHandler?
```

Comment:
https://github.com/eclipse/jetty.project/pull/5874 presumes a different scenario.

Reply:
Currently I am stumped here. @olamy any idea at all?

Reply:
> Currently I am stumped here.

@jglick I am disappointed you did not ask me. The instance of `ServletContextHandler` in Winstone is the `org.eclipse.jetty.webapp.WebAppContext` created in `winstone.HostConfiguration#create`. The call to `org.eclipse.jetty.websocket.server.config.JettyWebSocketServletContainerInitializer#configure` has to be done before the call to `org.eclipse.jetty.server.Server#start` so we might as well do it immediately after creating the `org.eclipse.jetty.webapp.WebAppContext` in `winstone.HostConfiguration#create`. With that in place, the call to `org.eclipse.jetty.server.Server#start` will do two important things: invoke `org.eclipse.jetty.websocket.server.config.JettyWebSocketConfiguration`'s constructor to expose the relevant classes, and invoke `org.eclipse.jetty.websocket.server.JettyWebSocketServerContainer#ensureContainer`, which creates the container and stashes it in the servlet context with attribute name `WebSocketContainer.class.getName()`, from which it can be fished out in the application.

So to summarize, patch Winstone with

```java
diff --git a/src/main/java/winstone/HostConfiguration.java b/src/main/java/winstone/HostConfiguration.java
index be7a223..21e9dab 100644
--- a/src/main/java/winstone/HostConfiguration.java
+++ b/src/main/java/winstone/HostConfiguration.java
@@ -14,6 +14,7 @@ import org.eclipse.jetty.server.Server;
 import org.eclipse.jetty.server.handler.ContextHandlerCollection;
 import org.eclipse.jetty.server.handler.RequestLogHandler;
 import org.eclipse.jetty.webapp.WebAppContext;
+import org.eclipse.jetty.websocket.server.config.JettyWebSocketServletContainerInitializer;
 import winstone.cmdline.Option;
 
 import javax.servlet.SessionTrackingMode;
@@ -181,6 +182,7 @@ public class HostConfiguration {
                 getSessionHandler().getSessionCache().setEvictionPolicy( sessionEviction );
             }
         };
+        JettyWebSocketServletContainerInitializer.configure(wac, null);
         wac.getSecurityHandler().setLoginService(loginService);
         wac.setThrowUnavailableOnStartupException(true);    // if boot fails, abort the process instead of letting empty Jetty run
         wac.setMimeTypes(mimeTypes);
```

and then amend this PR with

```java
diff --git a/websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java b/websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java
index af8b6125a1..694b9f6297 100644
--- a/websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java
+++ b/websocket/jetty10/src/main/java/jenkins/websocket/Jetty10Provider.java
@@ -33,6 +33,7 @@ import java.util.concurrent.Future;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.api.WebSocketContainer;
 import org.eclipse.jetty.websocket.api.WebSocketListener;
 import org.eclipse.jetty.websocket.api.WriteCallback;
 import org.eclipse.jetty.websocket.server.JettyServerUpgradeRequest;
@@ -59,11 +60,7 @@ public class Jetty10Provider implements Provider {
 
     private synchronized void init(HttpServletRequest req) throws Exception {
         if (container == null) {
-            // TODO JettyWebSocketServletContainerInitializer.configure(context, null) but where do we get the ServletContextHandler?
-            container = JettyWebSocketServerContainer.getContainer(req.getServletContext());
-            if (container == null) {
-                throw new IllegalStateException("not initialized");
-            }
+            container = (JettyWebSocketServerContainer) req.getServletContext().getAttribute(WebSocketContainer.class.getName());
         }
     }
```

I just tested this and it seems to work fine.

Reply:
BTW not sure if anybody has any strong feelings about whether that call to `JettyWebSocketServletContainerInitializer.configure` in Winstone should be done unconditionally or should be opt-in/opt-out via CLI and/or system property. If opt-in is desired, Jenkins could opt-in via `executable.Main` in the `war/`module in core (previously the `extras-executable-war` repository). I doubt anyone would ever have a need to turn this off, so my initial take is that it should just be enabled unconditionally for simplicity, but I do not feel strongly one way or another.

Reply:
(See also https://github.com/jenkinsci/jenkins-test-harness/pull/453/commits/7a2234286b681c5c4831295ef992a332e43e5f97)

Reply:
Many thanks! I must have been tired and did not think to look for _subtypes_ of `ServletContextHandler`.